### PR TITLE
gh-94635: Add Reference, How-to, and Concepts headings to sqlite3 docs

### LINKED ISSUE #94635: Restructure the sqlite3 docs table-of-contents
The current docs are laid out more or less like this:

- short introduction
- relatively short tutorial
- reference
- how-to's/guides and some in-depth explanations

Suggesting to make the following improvements:

- add top-level headings for:
    - introduction
    - tutorial
    - reference
    - guides
    - concepts
- reorder sections according to the new ToC
- add anchors for each top-level heading

Implemented by (all backported to 3.11 and 3.10):
- #95269 
- #94636 
- #95751
- #96089
- #94677 (closed, as it was implemented by other PRs)
- #96136

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-94635

### CODE REVIEW

File: Doc/library/sqlite3.rst [hunk 1]
```diff
@@ -1437,55 +1379,42 @@ This section shows recipes for common adapters and converters.
...
---------------------------------
 
+.. _sqlite3-shortcut-methods:
 
 Using shortcut methods
-^^^^^^^^^^^^^^^^^^^^^^
+~~~~~~~~~~~~~~~~~~~~~~
 
 Using the nonstandard :meth:`execute`, :meth:`executemany` and
 :meth:`executescript` methods of the :class:`Connection` object, your code can
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
Using the non-standard :meth:`~Connection.execute`,
:meth:`~Connection.executemany` and :meth:`~Connection.executescript`
methods of the :class:`Connection` class, your code can
```

Not directly related, but since I can reach it...

File: Doc/library/sqlite3.rst [hunk 2]
```diff
@@ -1362,40 +1334,10 @@ The following example illustrates the implicit and explicit approaches:
...
-   always returns a naive :class:`datetime.datetime` object. To preserve UTC
-   offsets in timestamps, either leave converters disabled, or register an
-   offset-aware converter with :func:`register_converter`.
-
-
 .. _sqlite3-adapter-converter-recipes:
 
 Adapter and Converter Recipes
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: IMO, particularly when framed as Guides/How-tos, it might make more sense to split the adapter and converter recipes/defaults and move them under their respective task-oriented "Using adapters" and "Converting SQLite values" to connect them to the high-level task and make each top-level heading correspond to the task they describe.

File: Doc/library/sqlite3.rst [hunk 3]
```diff
@@ -1535,6 +1466,103 @@ the context manager is a no-op.
...
+SQLite natively supports the following types: ``NULL``, ``INTEGER``,
+``REAL``, ``TEXT``, ``BLOB``.
+
+The following Python types can thus be sent to SQLite without any problem:
+
++-------------------------------+-------------+
+| Python type                   | SQLite type |
++===============================+=============+
+| :const:`None`                 | ``NULL``    |
++-------------------------------+-------------+
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: Arguably, this may be more reference material than an explanation/concepts, since it simply states the Python data types supported by the `sqlite3` module and what SQLite types they are converted to in a highly succinct, structured and information-oriented way, per the definition of Reference material:

> Reference guides are technical descriptions of the machinery and how to operate it. Reference material is information-oriented.

> The only purpose of a reference guide is to describe, as succinctly as possible, and in an orderly way.

> Reference material should be austere and to the point. One hardly reads reference material; one consults it.

and contains little to no high-level discussion, as is emblematic of explanation/concepts:

> Explanation is discussion that clarifies and illuminates a particular topic. Explanation is understanding-oriented.

> It’s not a close-up view of the machinery, like reference material. It’s documentation that approaches a topic from a higher perspective, and from different angles.

> This allows explanation to become discussion, a more relaxed, freer way to consider something. Explanation joins things together. It’s documentation that it makes sense to read while away from the product itself.

On the other hand, this does mean that there's only one top-level item in this category, and perhaps this section could be expanded with more explanation and background of the various types.

File: Doc/library/sqlite3.rst [hunk 4]
```diff
@@ -1535,6 +1466,103 @@ the context manager is a no-op.
...
+store additional Python types in an SQLite database via
+:ref:`object adapters <sqlite3-adapters>`,
+and you can let the :mod:`sqlite3` module convert SQLite types to
+Python types via :ref:`converters <sqlite3-converters>`.
+
+
+.. _sqlite3-controlling-transactions:
+
+Controlling Transactions
+~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
Transaction Control
~~~~~~~~~~~~~~~~~~~
```

Concept-oriented rather than task-oriented, reflecting both the Diataxis content type and the actual focus of the section

**Review by CAM-Gerlach — COMMENTED** [review]
Looks good overall; just a few comments.

File: Doc/library/sqlite3.rst (hunk 1 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Oh, there's room for improvement there... I'll save that for another PR. And when we touch it, the word "nonstandard" goes. (I want those completely purged from the sqlite3 docs.)

**Comment by erlend-aasland** [comment]
Thanks for the review, CAM. My first impression when reading the review comments is that this PR is premature; we want to improve the wording in headings, and possibly move around some sections in separate PRs first.

File: Doc/library/sqlite3.rst [hunk 5]
```diff
@@ -21,6 +25,12 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
 compliant with the DB-API 2.0 specification described by :pep:`249`, and
 requires SQLite 3.7.15 or newer.
 
+
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst:37]: Do you think it would be useful to have a ToC or even just a sentence with direct links to the main sections?

Something like:
```rst
If you are new to sqlite3, read the :ref:`sqlite3-tutorial`; for information about the objects
and their methods, skip to the :ref:`sqlite3-reference`; if you are trying to solve a particular
issue, check the :ref:`sqlite3-howtos`;  and for more theoretical background on the module,
see the :ref:`sqlites-concepts`.
```

or

```rst
This documented includes four main sections:
* :ref:`sqlite3-tutorial`: teaches how to use the sqlite3 module;
* :ref:`sqlite3-reference`: describes the classes and methods defined in the module;
* :ref:`sqlite3-howtos`: explains how to solve specific problems;
* :ref:`sqlites-concepts`: provides more background and insights about the design of the module;
```

(We might want to standardize this as we start adopting diataxis more widely.)

File: Doc/library/sqlite3.rst [hunk 6]
```diff
@@ -1230,60 +1245,15 @@ The exception hierarchy is defined by the DB-API 2.0 (:pep:`249`).
...
-you can let the :mod:`sqlite3` module convert SQLite types to different Python
-types via converters.
+Guides
+------
 
+.. _sqlite3-adapters:
 
 Using adapters to store custom Python types in SQLite databases
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: FWIW https://devguide.python.org/documenting/#sections recommends `^^^` after `---`.  I'm not sure how consistent its use is in the docs, and it doesn't matter too much.

**Review by ezio-melotti — COMMENTED** [review]
A step in the right direction!

File: Doc/library/sqlite3.rst (hunk 5 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:37]: Definitely; IMO the latter is the best option.

File: Doc/library/sqlite3.rst (hunk 6 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Right. Where does `~` belong?

File: Doc/library/sqlite3.rst (hunk 3 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > Arguably, this may be more reference material than an explanation/concepts [...]

Yeah, you are right. This is more of a reference than an explanation.

> On the other hand, this does mean that there's only one top-level item in this category, and perhaps this section could be expanded with more explanation and background of the various types.

That should not be a problem :)

File: Doc/library/sqlite3.rst (hunk 6 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: It isn't one of the six heading levels defined there (and going deeper than that is likely ill advised)—not the most important, but it does make it easier to move content between documents and keep track of what heading level you are at in the source.

File: Doc/library/sqlite3.rst (hunk 2 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I'm not sure about this. This implies we need to split the "default adapters/converters" section as well. They are both very small, so I'm not sure of there is added value for the user.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I'll keep it like it is for this PR. We can always split it up later. Marking this as resolved for this PR.

**Comment by erlend-aasland** [comment]
All right, I think this is ready for another round of reviews. I've already added the Introduction and Tutorial headings in a separate PR. For this PR, I restrict it to:

- Adding top-level Reference, How-to guides, and Concepts headings, and simply rearrange the existing subsections below these
- Normalise heading capitalisation.

Saved for later:
- Possibly reword headings (the other draft PR)
- Possibly split adapter/converter recipes section

File: Doc/library/sqlite3.rst (hunk 1 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I created #95334

File: Doc/library/sqlite3.rst [hunk 7]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
 compliant with the DB-API 2.0 specification described by :pep:`249`, and
 requires SQLite 3.7.15 or newer.
 
+This document includes four main sections:
+
+* :ref:`sqlite3-tutorial`: teaches how to use the sqlite3 module
+* :ref:`sqlite3-reference`: describes the classes and methods defined in the
+  module
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-reference` describes the classes and functions
  this module defines
```

* "methods" -> "functions" since "classes" include "methods" whereas much of the reference docs cover top-level functions
* Reduce wordy repetition
* Phrase as prose sentence

File: Doc/library/sqlite3.rst [hunk 8]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
 compliant with the DB-API 2.0 specification described by :pep:`249`, and
 requires SQLite 3.7.15 or newer.
 
+This document includes four main sections:
+
+* :ref:`sqlite3-tutorial`: teaches how to use the sqlite3 module
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-tutorial` teaches how to use the sqlite3 module
```

IMO, it reads better and sounds a little friendlier if these are phrased as prose sentences

File: Doc/library/sqlite3.rst [hunk 9]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
...
+* :ref:`sqlite3-reference`: describes the classes and methods defined in the
+  module
+* :ref:`sqlite3-howtos`: explains how to solve specific problems
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-howtos` explains how to handle specific tasks
```

* Tweak wording to mention positive (and more general) "tasks" rather than negative (and not always encompassing) "problems"
* Frame as a prose sentence

File: Doc/library/sqlite3.rst [hunk 10]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
...
+* :ref:`sqlite3-concepts`: provides more background and insights about the
+  design of the module
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-concepts` provides more background and insights
  about the module's design
```

* Reword to flow better and be a bit more concise
* Rephrase as a sentence

File: Doc/library/sqlite3.rst (hunk 9 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: The singular/plural agreement sounds kinda weird, even if it is technically correct (referring to the singular section rather than the plural guides). I considered proposing referring to the underlying item rather than the section name to avoid this, but decided not to for now. We _could_ reference the guides rather than the section here, but that wouldn't be consistent with the others (especially since Concepts is also a plural word, even if the object it references, a section, is singular, whereas the "guides" are actually multiple distinct objects, just under one section)

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Or maybe I'm overthinking this, heh

File: Doc/library/sqlite3.rst [hunk 11]
```diff
@@ -1666,6 +1611,85 @@ the context manager is a no-op.
 .. literalinclude:: ../includes/sqlite3/ctx_manager.py
 
 
+.. _sqlite3-uri-tricks:
+
+SQLite URI tricks
+^^^^^^^^^^^^^^^^^
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
Working with SQLite URIs
^^^^^^^^^^^^^^^^^^^^^^^^
```

Frame this as a how-to

**Review by CAM-Gerlach — COMMENTED** [review]
A few minor textual suggestions, otherwise LGTM, thanks

File: Doc/library/sqlite3.rst (hunk 9 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: > We could reference the guides rather than the section here [...]

I don't think that is needed, as you've got the ToC menu to your immediate left ;)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I do like the proposed wording, though.

File: Doc/library/sqlite3.rst (hunk 10 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Actually, I'd like to reword this completely, because it does _not_ talk about the module's design. The only item in "Concepts" is the transaction control explanation. The next upcoming item is the new PEP 249 autocommit transaction control explanation. So, two explanations; no concepts 😆

File: Doc/library/sqlite3.rst (hunk 9 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: > I don't think that is needed, as you've got the ToC menu to your immediate left ;)

Sorry, I didn't mean to imply actually enumerating the guides or anything, but rather the semantic nuance of referring to "How-to guides" the _section_ (singular) versus "How-to guides" the _guides_ (plural)

File: Doc/library/sqlite3.rst (hunk 10 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: PTAL

File: Doc/library/sqlite3.rst (hunk 9 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Ah, got it :)

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I'm marking this as resolved. Let me know if you disagree :)

File: Doc/library/sqlite3.rst [hunk 12]
```diff
@@ -1666,6 +1611,85 @@ the context manager is a no-op.
...
+More information about this feature, including a list of parameters,
+can be found in the `SQLite URI documentation`_.
+
+.. _SQLite URI documentation: https://www.sqlite.org/uri.html
+
+
+.. _sqlite3-explanation:
+
+Concepts
+--------
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: Did you mean to change the actual heading as well?

```suggestion
Explanation
-----------
```

File: Doc/library/sqlite3.rst [hunk 13]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
 compliant with the DB-API 2.0 specification described by :pep:`249`, and
 requires SQLite 3.7.15 or newer.
 
+This document includes four main sections:
+
+* :ref:`sqlite3-tutorial` teaches how to use the sqlite3 module.
+* :ref:`sqlite3-reference` describes the classes and functions this module
+  defines.
+* :ref:`sqlite3-howtos` explains how to handle specific tasks.
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-howtos` details how to handle specific tasks.
```

Avoid ambiguity/confusing in using "explain" for a non-explanation section (didn't catch this the first time, sorry)

File: Doc/library/sqlite3.rst [hunk 14]
```diff
@@ -25,6 +25,15 @@ The sqlite3 module was written by Gerhard Häring.  It provides an SQL interface
...
+* :ref:`sqlite3-explanation` provides in-depth understanding of transaction
+  control.
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
* :ref:`sqlite3-explanation` provides in-depth background on
  transaction control.
```

Fix grammar and phrasing issues

**Review by CAM-Gerlach — COMMENTED** [review]
Just a couple minor followups, otherwise LGTM, thanks

**Comment by erlend-aasland** [comment]
Thanks for the reviews! I'll let this sit around for some more days before merging, to give @ezio-melotti a chance to review the final changes.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM on my end, thanks @erlend-aasland

File: Doc/library/sqlite3.rst [hunk 15]
```diff
@@ -1369,10 +1379,18 @@ This is how SQLite types are converted to Python types by default:
 +-------------+----------------------------------------------+
 
 The type system of the :mod:`sqlite3` module is extensible in two ways: you can
-store additional Python types in an SQLite database via object adaptation, and
-you can let the :mod:`sqlite3` module convert SQLite types to different Python
-types via converters.
+store additional Python types in an SQLite database via
+:ref:`object adapters <sqlite3-adapters>`,
+and you can let the :mod:`sqlite3` module convert SQLite types to
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst]: Just a minor nit: the module is already linked earlier in the sentence, so this second occurrence could be converted to ` ``sqlite3`` `.  I'm not sure if this convention is followed throughout the document (I saw at least another ` ``sqlite3`` `), and the style guide doesn't seem to mention this.

If the two versions render differently, it might be better to use `:mod:` or just ` ``sqlite3`` ` for both, given that in this case the links points to the document itself and they are not particularly useful.

File: Doc/library/sqlite3.rst [hunk 16]
```diff
@@ -1666,6 +1611,85 @@ the context manager is a no-op.
...
+.. _sqlite3-controlling-transactions:
+
+Transaction control
+^^^^^^^^^^^^^^^^^^^
+
+The ``sqlite3`` module does not adhere to the transaction handling recommended
+by :pep:`249`.
+
+If the connection attribute :attr:`~Connection.isolation_level`
+is not :const:`None`,
```
Comment (ezio-melotti) [inline Doc/library/sqlite3.rst:1714]: Not your fault since you are just moving an existing, but for `None`, ` ``None`` ` is preferred.  This might belong in a different PR though, assuming you think it's worth fixing it.

**Review by ezio-melotti — APPROVED** [review]

File: Doc/library/sqlite3.rst (hunk 15 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: I was going to point this out too; it just slipped my mind for some reason. It seems the convention @erlend-aasland has followed and advocated for (as discussed in the review feedback of one of his other recent PRs) is to just always use regular literals for the module name inside the same module.

The rendering for both (at least under the current theme) is as a literal, just with the `:mod:` role version also being a link (which isn't necessary here), though at least for this particular case it is moot since the convention seems to be just going with a verbatim literal including on first use.

File: Doc/library/sqlite3.rst (hunk 16 above)
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:1714]: To note, I don't have a strong opinion, but ``:const:`None` `` is the general convention that @erlend-aasland has used so far through the revamped ``sqlite3`` docs, so it might be better to make this change separately if so. Is there somewhere this preference is documented?

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: Oh, I was not aware of that convention; I've been using the reference version consistently in all my PRs. Is that also the convention for True and False?

File: Doc/library/sqlite3.rst (hunk 15 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]:
```suggestion
and you can let the ``sqlite3`` module convert SQLite types to
```

File: Doc/library/sqlite3.rst (hunk 16 above)
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: Changing all those in sqlite3.rst would be a large diff :) _If_ I were to do that, I prefer to do it in a separate PR.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: I'm going to mark this as resolved, as I won't be making any changes in this PR. I'll wait a little bit, in case other comments appear, though.

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:1714]: Seems like its discussed in the [devguide](https://devguide.python.org/documentation/markup/#rest-inline-markup), but under the "reStructuredText Markup" page rather than the style guide, which mostly repeats general details of reST/Sphinx, and under the ``keyword`` role rather than the ``const`` role in which they are actually indexed under

image

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: Ok, makes sense. Thanks for finding this.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: We should probably move that to the style guide.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1714]: I consider it low priority, though :)

**Comment by erlend-aasland** [comment]
Thanks for the reviews, CAM and Ezio!

File: Doc/library/sqlite3.rst (hunk 16 above)
Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:1714]: > Is that also the convention for True and False?

See also https://github.com/python/devguide/pull/916, where I tried to make this more explicit.

File: Doc/library/sqlite3.rst [hunk 17]
```diff
@@ -1610,13 +1610,14 @@ This section shows recipes for common adapters and converters.
    sqlite3.register_converter("timestamp", convert_timestamp)
 
 
-.. _sqlite3-shortcut-methods:
+.. _sqlite3-connection-shortcuts:
```
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst:1613]: Just a general comment—in the future, it would be best to just leave the existing ref target in place (either instead of or addition to adding a new one), since not doing so moots much of the point of using ref targets in the first place—it requires unnecessarily changing all the existing inbound links, has a high chance of breaking some regardless (in a large doc repo like CPython that doesn't use `-n`), breaks any inbounds by external repos using intersphinx, and also breaks any normal hyperlinks linked to the `ref` target `<span>` instead of the page heading (e.g. via copying an internal inbound link).

Seems like it would be worth me opening a docs community issue or just a PR directly to the devguide to mention this (which it doesn't)? To note, the devguide currently also doesn't mention (in either the style guide or the reST Markup page under [Cross Links](https://devguide.python.org/documentation/markup/#doc-ref-role)) how ref targets should be "namespaced", i.e. by the module or page name, which is more generally useful than just repeating the basics of how to use them already covered in the Sphinx quick reference or the reST docs.

@ezio-melotti thoughts?

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1613]: Yes, sorry 'bout that :(

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1613]: I'm not sure it is worth reverting that particular change; that anchor was introduced very recently.

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:1613]: Fair enough, that is an exception—though I'm not sure if it was used somewhere in a NEWS entry, etc.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:1613]: I'll have a look! **UPDATE**: Seems not. Anyway, I'll be more careful in the future. Thanks for the heads-up.

Comment (ezio-melotti) [reply Doc/library/sqlite3.rst:1613]: @CAM-Gerlach:
> @ezio-melotti thoughts?

A bigger problem is that AFAIU changing ref targets also breaks anchors.  I.e., https://docs.python.org/3.12/library/sqlite3.html#sqlite3-shortcut-methods no longer works after the ref name has been changed.  For this specific case it's probably not an issue, but it's something to keep in mind when updating/moving references.

Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst:1613]: I opened python/devguide#927 to document this
